shared/src/types/hash.rs

```diff
@@ -145,3 +154,80 @@ impl Value for Hash {
         Hash([0u8; 32])
     }
 }
+
+/// A hex encoded hash.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct HexEncodedHash {
```

Comment:
It looks like your branch doesn't have the `From<&str>` methods for hash yet, but they are coming in my PRs. Since `Display` on `Hash` is hex, the from string method will do all of this. So I'm not sure you need a special type here. In fact, I don't see why you couldn't just use a `&str`

Reply:
the main motivation to add `HexEncodedHash` was to parse hex encoded hashes `$tx_hash` we receive in `/<accepted|applied>/<$tx_hash>` into a stack allocated value. we know these hashes can only of 64 bytes, so we don't need to heap-alloc a `String` for them.

Reply:
I also personally prefer small domain specific types, which uphold some invariants

Reply:
https://github.com/anoma/namada/blob/93000b0d28c9ed54b3a597b9cfddc7a99adda6bf/shared/src/types/hash.rs#L101

Reply:
That code should work. It does pass through a heap allocation, so it could be adjusted to use a `try_into` to go directlu into a 32 byte array.

Reply:
Anyway, the motivation is clear, I'm just trying to prevent double writing code in two separate PRs. So I'm suggesting you steal some upcoming changes.

Reply:
> I also personally prefer small domain specific types, which uphold some invariants

But our `Hash` type should be hex encoded as an invariant. Is there some other invariant I am unaware of?

Reply:
not really, I don't think. the bytes in the hash itself are completely arbitrary.